gh-108948: tarfile should handle sticky bit in FreeBSD

sorcio opened the pull request:
Instead of failing on FreeBSD, we emulate the behavior of other platforms. That means that if a file wants to set the sticky bit and we are not root, we just ignore it.


<!-- gh-issue-number: gh-108948 -->
* Issue: gh-108948
<!-- /gh-issue-number -->

sorcio committed:
tarfile: fix sticky bit handling in FreeBSD

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

sorcio committed:
fix blurb

sorcio committed:
fix blurb

sorcio committed:
fix chmod exception handler in tarfile

vstinner commented on Lib/tarfile.py:
```diff
@@ -2569,6 +2570,13 @@ def chmod(self, tarinfo, targetpath):
+                # On FreeBSD, chmod fails when trying to set the sticky bit on
+                # a file as a normal user. It's a noop in most other platforms.
+                try:
+                    os.chmod(targetpath, tarinfo.mode & ~stat.S_ISVTX)
```
if chmod() succeed, you will raise ExtractError. Is it on purpose?

Maybe put your code code inside the try block, so you don't have to copy/paste the except/raise block?

sorcio replied:
Uh, no, it's very much not on purpose. I changed it. It's not the same exception exactly. The user will never see the OSError with EFTYPE. A new one could be raised on the second invocation. And ExtractError is raised `from e`. I wanted to make that obvious from the code but maybe it's not!

vstinner reviewed (commented):
I confirm that test_tarfile.test_modes() fails on FreeBSD without this change, and it does pass with this change.

But I don't think that the tarfile change is correct.

sorcio committed:
chain exceptions

vstinner reviewed (commented):
I would prefer to keep the first exception and ignore the second chmod() exception:

```py
    def chmod(self, tarinfo, targetpath):
        """Set file permissions of targetpath according to tarinfo.
        """
        if tarinfo.mode is None:
            return

        try:
            try:
                os.chmod(targetpath, tarinfo.mode)
            except OSError as exc:
                if not(hasattr(errno, "EFTYPE") and e.errno == errno.EFTYPE):
                    raise

                # gh-108948: On FreeBSD, chmod() fails when trying to set the
                # sticky bit on a file as a normal user. It's a noop in most
                # other platforms.
                try:
                    os.chmod(targetpath, tarinfo.mode & ~stat.S_ISVTX)
                except OSError:
                    # re-raise the first chmod() exception,
                    # ignore the second chmod() exception.
                    raise exc from None
                else:
                    # second chmod() succeeded
                    pass
        except OSError:
            raise ExtractError("could not change mode")
```

Enjoy the 3 levels of nested try/exception :-)

sorcio committed:
test exception branches

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3818,7 +3845,15 @@ def test_modes(self):
+            os.chmod(tmp_filename, new_mode)
+        except OSError as err:
+            # FreeBSD fails with EFTYPE if sticky bit cannot be set, instead
+            # of ignoring it.
```
```suggestion
            # gh-108948: FreeBSD fails with EFTYPE if sticky bit cannot be set,
            # instead of ignoring it.
```

sorcio committed:
Update Lib/test/test_tarfile.py

Co-authored-by: Victor Stinner <vstinner@python.org>

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -690,6 +691,32 @@ def format_mtime(mtime):
+                other_error = OSError(errno.EPERM, "different error")
+                mock.side_effect = [eftype_error, other_error]
+                with self.assertRaises(tarfile.ExtractError) as excinfo:
+                    tar.extract("sticky", DIR)
```
Can you please test somewhere that the ExtractError error message is "could not change mode"?

Can you use a different name? extract("sticky", DIR) could also fail because the file already exists. Maybe add "sticky2" to the archive.

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -690,6 +691,32 @@ def format_mtime(mtime):
+            arc.add("sticky", mode="?rw-rw-rwt")
+        tar = arc.open(errorlevel=2)
+        DIR = os.path.join(TEMPDIR, "chmod")
+        os.mkdir(DIR)
```
nitpick: you can add the mkdir() inside the try block.

vstinner commented on Lib/tarfile.py:
```diff
@@ -2567,7 +2568,19 @@ def chmod(self, tarinfo, targetpath):
+
+                # gh-108948: On FreeBSD, chmod() fails when trying to set the
+                # sticky bit on a file as non-root. But it's a noop in most
+                # other platforms, and we try and match that behavior here.
```
Do you mean that on other platforms chmod() silently ignores the sticky bit for non-root users? I'm not sure that I understood correctly.

sorcio replied:
Yes, exactly. I will reword more clearly. I was being stingy on the number of lines to take for this comment :) Same in the similar comment in the test case. I will reword there too.

vstinner commented on Lib/tarfile.py:
```diff
@@ -2567,7 +2568,19 @@ def chmod(self, tarinfo, targetpath):
+                # sticky bit on a file as non-root. But it's a noop in most
+                # other platforms, and we try and match that behavior here.
+                try:
+                    os.chmod(targetpath, tarinfo.mode & ~stat.S_ISVTX)
```
```suggestion
                    # Retry without the sticky bit
                    os.chmod(targetpath, tarinfo.mode & ~stat.S_ISVTX)
```

vstinner commented on Lib/tarfile.py:
```diff
@@ -2567,7 +2568,19 @@ def chmod(self, tarinfo, targetpath):
+            try:
+                os.chmod(targetpath, tarinfo.mode)
+            except OSError as e:
+                if not(hasattr(errno, "EFTYPE") and e.errno == errno.EFTYPE):
```
Maybe only retry chmod() when it's needed? Or is this test redundant with checking for EFTYPE?

```suggestion
                if not(hasattr(errno, "EFTYPE") and e.errno == errno.EFTYPE and (tarinfo.mode & stat.S_ISVTX)):
```

sorcio replied:
Yes, at least on FreeBSD. The documentation is very explicit about the condition causing the error:

> The effective user ID is not the super-user, the mode includes the sticky bit (S_ISVTX), and path does not refer to a directory.

I don't know of other systems with the same error, so the hasattr() check might be enough. But it might be a good idea to be defensive. It's not like it's an expensive check.

vstinner replied:
Ok, no need to test for the bit in this case.

vstinner commented on Lib/tarfile.py:
```diff
@@ -2567,7 +2568,19 @@ def chmod(self, tarinfo, targetpath):
-            os.chmod(targetpath, tarinfo.mode)
+            try:
+                os.chmod(targetpath, tarinfo.mode)
+            except OSError as e:
```
Please use variable names longer than 1 character: you can use exc and exc2.

sorcio replied:
I'm in favor of this but let me double check: throughout the file the convention seems to be to use `except ... as e:`. Is it okay to break the convention?

vstinner replied:
Please break the convention, it's a bad convention :-)

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -690,6 +691,32 @@ def format_mtime(mtime):
+        os.mkdir(DIR)
+        try:
+            # this should not raise:
+            tar.extract("sticky", DIR)
```
Can you check permissions of the created file? Please add also an empty line for readability.

sorcio replied:
> Can you check permissions of the created file?

Do you mean to stat the file and add an assertion that the flags are as expected? Sure, it's a good idea. The result would depend on whether the test code is run as root or not. Would it be fair to skip if running as root?

vstinner replied:
I expect that if run as root, the test sets the sticky bit. Otherwise, it doesn't. Python has a Cirrus CI job which runs the Python test suite on FreeBSD... as root!

sorcio replied:
Ok, already did it in the new commit.

Besides, pretty nice to have at least one test runner as root! (Well, assuming it works out with Cirrus credits)

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -690,6 +691,32 @@ def format_mtime(mtime):
+            self.assertEqual(excinfo.exception.__cause__.__cause__, eftype_error)
+        finally:
+            os_helper.rmtree(DIR)
+            tar.close()
```
Instead, you can use ``with tar:`` and ``self.addCleanup(os_helper.rmtree, DIR)`` above.

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3818,7 +3845,15 @@ def test_modes(self):
+            # gh-108948: FreeBSD fails with EFTYPE if sticky bit cannot be set,
+            # instead of ignoring it.
+            if hasattr(errno, "EFTYPE") and err.errno == errno.EFTYPE:
+                os.chmod(tmp_filename, new_mode & ~stat.S_ISVTX)
```
```suggestion
                # Retry without the sticky bit
                os.chmod(tmp_filename, new_mode & ~stat.S_ISVTX)
```

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3818,7 +3845,15 @@ def test_modes(self):
+            os.chmod(tmp_filename, new_mode)
+        except OSError as err:
+            # gh-108948: FreeBSD fails with EFTYPE if sticky bit cannot be set,
+            # instead of ignoring it.
```
"instead of ignoring it" is not clear. Maybe just remove it? Or explain that other platforms behave differently? You may mention that this is the behavior for non-root users.

sorcio committed:
add explanatory comment

Co-authored-by: Victor Stinner <vstinner@python.org>

sorcio committed:
apply review suggestions

sorcio committed:
fix expected mode

sorcio committed:
avoid deprecation warnings

vstinner commented on Lib/tarfile.py:
```diff
@@ -2567,7 +2568,26 @@ def chmod(self, tarinfo, targetpath):
+                # sticky bit on a file as non-root. On other platforms, the bit
+                # is silently ignored if it cannot be set. We make FreeBSD
+                # behave like other platforms by catching the error and trying
+                # again without the sticky bit.
```
Thank you 🙏 Honestly, the new comment is now straightforward and there is no doubt about its meaning :-)

vstinner commented on Lib/tarfile.py:
```diff
@@ -2567,7 +2568,26 @@ def chmod(self, tarinfo, targetpath):
+                        # The error from the second attempt is the direct cause
+                        # of the ExtractError, but we keep the original error
+                        # around for good information.
+                        raise exc2 from exc1
```
Nested try/except are complicated. I suggest adding this comment for readability.

```suggestion
                        raise exc2 from exc1

                    # the second chmod() call without sticky bit succeeded
```

sorcio replied:
Done

vstinner commented on Lib/tarfile.py:
```diff
@@ -2567,7 +2568,26 @@ def chmod(self, tarinfo, targetpath):
+                else:
+                    raise
         except OSError as e:
             raise ExtractError("could not change mode") from e
```
```suggestion
        except OSError as exc3:
            raise ExtractError("could not change mode") from exc3
```

sorcio replied:
Done

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -690,6 +691,36 @@ def format_mtime(mtime):
             tar.close()
             os_helper.rmtree(DIR)
 
+    @unittest.skipUnless(hasattr(errno, "EFTYPE"), "errno.EFTYPE required")
```
Would it make sense to mock this constant to test the code on other platforms?

sorcio replied:
I decided to go with it. Maybe it's overkill for such a niche case, but more tests rarely hurt. All new tests never skip.

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -690,6 +691,36 @@ def format_mtime(mtime):
+        self.addCleanup(os_helper.rmtree, DIR)
+        with tar:
+            # this should not raise:
+            tar.extract("sticky1", DIR, filter="fully_trusted")
```
I think that you should split this test in two tests:

* a method which doesn't mock anything and jsut check that tarfile works as expected: this line. -- functional test
* another method which mocks chmod() to test the complicated EFTYPE code path.

sorcio replied:
Done. Again, maybe overkill, but it's three tests now.

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -690,6 +691,36 @@ def format_mtime(mtime):
+                other_error = OSError(errno.EPERM, "different error")
+                mock.side_effect = [eftype_error, other_error]
+                with self.assertRaises(tarfile.ExtractError) as excinfo:
+                    tar.extract("sticky2", DIR, filter="fully_trusted")
```
Since tarfile implementation now uses a complicated triple-nested try block, would it be possible to test also the othe code path: EFTYPE error but then the second chmod() succeed.

sorcio replied:
I think this is done now.

vstinner commented on Misc/NEWS.d/next/Library/2023-09-05-17-11-17.gh-issue-108948.HL1Ttw.rst:
```diff
@@ -0,0 +1 @@
+On FreeBSD, :mod:`tarfile` will not attempt to set the sticky bit on extracted files when it's not possible, matching the behavior of other platforms.
```
Hum. I suggest to explain it differently.

```
:mod:`tarfile`: On FreeBSD, if the sticky bit cannot be set,
if the user is not root for example, clear the sticky bit,
matching the behavior of other platforms.
```

vstinner replied:
Maybe it would be worth it to add a note about the sticky bit in https://docs.python.org/dev/library/tarfile.html#tarfile.TarFile.extract documentation?

sorcio replied:
I went with Ethan's suggestion for the blurb.

I also added a note, but under the filters documentation, so if a user searches for "sticky" or "bits" or similar keywords they find relevant information right there. I can change it or add more, of course.

vstinner reviewed (commented):
@encukou: Oh good, it seems like ``filter='data'`` already clears the sticky bit in _get_filtered_attrs():

```py
        # Strip high bits & group/other write bits
        mode = mode & 0o755
```

Is it a new functional test needed to ensure that the sticky bit is ignored with ``filter='data'``?

ethanfurman commented on Misc/NEWS.d/next/Library/2023-09-05-17-11-17.gh-issue-108948.HL1Ttw.rst:
```diff
@@ -0,0 +1 @@
+On FreeBSD, :mod:`tarfile` will not attempt to set the sticky bit on extracted files when it's not possible, matching the behavior of other platforms.
```
I find the current explanation much clearer.  To go with @vstinner's format:
```suggestion
:mod:`tarfile`:  On FreeBSD, ``tarfile`` will not attempt to set the sticky bit on extracted files when it's not possible, matching the behavior of other platforms.
```

sorcio committed:
move tarfile eftype test

test_extract_chmod_eftype was run once per testtar.* file but it did not use those files

sorcio committed:
eftype test should handle macos

sorcio committed:
use can_chmod_set_sticky helper where possible

sorcio committed:
bit of polishing

sorcio committed:
split and refactor tests

sorcio committed:
make TarFile.chmod slightly more readable

sorcio committed:
document tarfile behavior with sticky bit on files

sorcio committed:
Update Misc/NEWS.d/next/Library/2023-09-05-17-11-17.gh-issue-108948.HL1Ttw.rst

Co-authored-by: Ethan Furman <ethan@stoneleaf.us>

vstinner commented on Doc/library/tarfile.rst:
```diff
@@ -939,6 +939,13 @@ reused in custom filters:
+
+      Although the filter preserves all permission bits unchanged, the sticky
+      bit (:const:`~stat.S_ISVTX`) will not be set on extracted files when the
+      system does not allow it. For compatibility reasons, no error is raised
```
```suggestion
      system does not allow it for the current user. For compatibility reasons, no error is raised
```

sorcio replied:
Uh, yeah, it's right that running as root will set the sticky bit in the cases we are testing. But I don't want to suggest it's a permissions issue (which would likely raise EPERM), or to imply more than is necessary. E.g. I think file system implementations on Linux are allowed to silently drop the bit if they don't support it *regardless of the user*. What do you think is worth clarifying in this context?

vstinner replied:
You can ignore my suggestion if you think that it's irrelevant.

sorcio replied:
Rephrased ever so slightly for readability, but I avoided the "for the current user" part because it feels out of place.

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3056,6 +3057,94 @@ def test_keyword_only(self, mock_geteuid):
+                mode = os.stat(f.fileno()).st_mode
+                return bool(mode & stat.S_ISVTX)
+        finally:
+            os.unlink(filename)
```
It's safer to use os_helper.unlink() here.

sorcio replied:
Done.

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3056,6 +3057,94 @@ def test_keyword_only(self, mock_geteuid):
+
+
+def can_chmod_set_sticky():
+    """Can chmod set the sticky bit on a file?"""
```
Please explain in the docstring that it returns False if chmod() doesn't fail but ignores the sticky bit.

sorcio replied:
Done.

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3056,6 +3057,94 @@ def test_keyword_only(self, mock_geteuid):
+    def can_chmod_set_sticky_inner():
+        if not os_helper.can_chmod():
+            return False
+        filename = os_helper.TESTFN + "-chmod-sticky"
```
honestly, just use "chmod-sticky", to avoid further troubles with non-ASCII filenames.

sorcio replied:
Decided not to do it, and instead just use TESTFN like `can_chmod()` does. The concern is fair but I don't want to touch os_helper more than necessary.

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3056,6 +3057,94 @@ def test_keyword_only(self, mock_geteuid):
+_can_chmod_set_sticky = None
+
+
+def can_chmod_set_sticky():
```
You should move it to support.os_helper.

I suggest to rename to can_chmod_set_sticky_bit().

sorcio replied:
Done.

sorcio committed:
apply review suggestions

vstinner commented on Lib/test/support/os_helper.py:
```diff
@@ -275,6 +275,39 @@ def skip_unless_working_chmod(test):
+
+    global _can_chmod_set_sticky_bit
+    if _can_chmod_set_sticky_bit is None:
+        _can_chmod_set_sticky_bit = can_chmod_set_sticky_bit_inner()
```
Hum, I suggest you to mimick code around in os_helper.py: other "can" functions also have a cache (global variable), but they don't use an "inner" function.

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3056,6 +3057,69 @@ def test_keyword_only(self, mock_geteuid):
+            tar.extract(file_name, self.test_dir, filter="fully_trusted")
+        self.extracted_path = os.path.join(self.test_dir, file_name)
+
+    def test_extract_chmod_eftype_success(self):
```
On most platforms, this test willl be run without EFTYPE. I suggest to remove it:

```suggestion
    def test_extract_chmod_success(self):
```

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3056,6 +3057,69 @@ def test_keyword_only(self, mock_geteuid):
+        self.assertEqual(got_mode, expected_mode)
+
+    @unittest.mock.patch.object(errno, "EFTYPE", fake_EFTYPE, create=True)
+    def test_extract_chmod_eftype_mock(self):
```
```suggestion
    def test_extract_mock_chmod_eftype_success(self):
```

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3056,6 +3057,69 @@ def test_keyword_only(self, mock_geteuid):
+    @unittest.mock.patch.object(errno, "EFTYPE", fake_EFTYPE, create=True)
+    def test_extract_chmod_eftype_mock(self):
+        # Like test_extract_chmod_eftype_success but mock chmod so we can
+        # explicitly test the case we want, regardless of the OS.
```
I would prefer to describe which code path is tested:

Test chmod() failing with OSError(EFTYPE) and then pass when re-run without the sticky bit.

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3056,6 +3057,69 @@ def test_keyword_only(self, mock_geteuid):
+        self.extracted_path = os.path.join(self.test_dir, file_name)
+
+    def test_extract_chmod_eftype_success(self):
+        # Extracting a file as non-root should skip the sticky bit (gh-108948)
```
```suggestion
        # gh-108948: Extracting a file as non-root should skip the sticky bit
```

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3056,6 +3057,69 @@ def test_keyword_only(self, mock_geteuid):
+        self.assertEqual(chmod_mode1 | chmod_mode2, chmod_mode1)
+
+    @unittest.mock.patch.object(errno, "EFTYPE", fake_EFTYPE, create=True)
+    def test_extract_chmod_eftype_error(self):
```
```suggestion
    def test_extract_mock_chmod_eftype_error(self):
```

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3056,6 +3057,69 @@ def test_keyword_only(self, mock_geteuid):
+
+    @unittest.mock.patch.object(errno, "EFTYPE", fake_EFTYPE, create=True)
+    def test_extract_chmod_eftype_error(self):
+        # Take care that any other error is preserved in the EFTYPE case.
```
I suggest to add first (but preserve your longer comment):

Test chmod() failing with OSError(EFTYPE) and then raise PermissionError() when re-run without the sticky bit.

vstinner commented on Lib/test/test_tarfile.py:
```diff
@@ -3056,6 +3057,69 @@ def test_keyword_only(self, mock_geteuid):
+            with self.assertRaises(tarfile.ExtractError,
+                                   msg="could not change mode") as excinfo:
+                self.extract_sticky_file()
+        self.assertEqual(excinfo.exception.__cause__, other_error)
+        self.assertEqual(excinfo.exception.__cause__.__cause__, eftype_error)
```
```suggestion
        self.assertIs(excinfo.exception.__cause__, other_error)
        self.assertIs(excinfo.exception.__cause__.__cause__, eftype_error)
```

vstinner reviewed (commented):
Sorry, this PR took longer what I expected, since it's a core feature of tarfile, and a security sensitive topic. But I prefer to take my time :-) You enhanced the doc and the tests, that's great!

Here is I hope my last review :-) We should be good to merge soon.

vstinner reviewed (changes requested):
Apparently, I misunderstood the issue and the issue should be analyzed deeper to come with a better fix.
